File: CIS194_intro_haskell/2013_spring/hw1/hanoi.erl
-module(hanoi).

-include_lib("eunit/include/eunit.hrl").

-type peg() :: [integer()].
-type peg_name() :: a | b | c.
-type step() :: ab | ac | bc | ba | cb | ca.
-type steps() :: [step()].
-type state() :: {peg(), peg(), peg(), steps()}.

-export([init/1, solve/1]).


-spec init(integer()) -> state().
init(NumDisks) ->
    {lists:seq(1, NumDisks), [], [], []}.


-spec solve(state()) -> state().
solve(State) ->
    {A0, _, _, _} = State,
    {A, B, C, Steps} = move(length(A0), a, b, c, State),
    {A, B, C, lists:reverse(Steps)}.


-spec move(NumDisks :: integer(),
           From     :: peg_name(),
           To       :: peg_name(),
           Temp     :: peg_name(),
           State    :: state()) -> state().
move(0, _, _, _, State) -> State;
move(1, From, To, _, State) ->
    step(From, To, State);
move(2, From, To, Temp, State) ->
    State2 = step(From, Temp, State),
    State3 = step(From, To, State2),
    step(Temp, To, State3);
move(NumDisks, From, To, Temp, State) ->
    State2 = move(NumDisks-1, From, Temp, To, State),
    State3 = step(From, To, State2),
    move(NumDisks-1, Temp, To, From, State3).


-spec step(peg_name(), peg_name(), state()) -> state().
step(FromPeg, ToPeg, {A, B, C, Steps}) ->
    case {FromPeg, ToPeg} of
        {a, b} -> {F, T} = place(A, B), {F, T, C, [ab | Steps]};
        {a, c} -> {F, T} = place(A, C), {F, B, T, [ac | Steps]};
        {b, c} -> {F, T} = place(B, C), {A, F, T, [bc | Steps]};
        {b, a} -> {F, T} = place(B, A), {T, F, C, [ba | Steps]};
        {c, b} -> {F, T} = place(C, B), {A, T, F, [cb | Steps]};
        {c, a} -> {F, T} = place(C, A), {T, B, F, [ca | Steps]}
    end.


-spec place(FromPeg :: peg(), ToPeg :: peg()) -> {peg(), peg()}.
place([Top | Rest], []) -> {Rest, [Top]};
place([Top | Rest], [H | _] = ToPeg) when Top < H -> {Rest, [Top | ToPeg]}.


solve_test() ->
    S0 = hanoi:init(0),
    ?assertEqual({[],[],[],[]}, S0),
    ?assertEqual({[],[],[],[]}, hanoi:solve(S0)),

    S1 = hanoi:init(1),
    ?assertEqual({[1],[],[],[]}, S1),
    ?assertEqual({[],[1],[],[ab]}, hanoi:solve(S1)),

    S2 = hanoi:init(2),
    ?assertEqual({[1,2],[],[],[]}, S2),
    ?assertEqual({[],[1,2],[],[ac,ab,cb]}, hanoi:solve(S2)),

    S3 = hanoi:init(3),
    ?assertEqual({[1,2,3],[],[],[]}, S3),
    ?assertEqual({[],[1,2,3],[],[ab,ac,bc,ab,ca,cb,ab]}, hanoi:solve(S3)),

    S4 = hanoi:init(4),
    ?assertEqual({[1,2,3,4],[],[],[]}, S4),
    ?assertEqual({[], [1,2,3,4], [],
                  [ac,ab,cb,ac,ba,bc,ac,ab,cb,ca,ba,cb,ac,ab,cb]},
                 hanoi:solve(S4)),

    S5 = hanoi:init(5),
    ?assertEqual({[1,2,3,4,5],[],[],[]}, S5),
    ?assertEqual({[],[1,2,3,4,5],[],
                  [ab,ac,bc,ab,ca,cb,ab,ac,bc,ba,ca,bc,ab,ac,bc,ab,ca,cb,ab,ca,bc,ba,ca,cb,ab,ac,bc,ab,ca,cb,ab]},
                 hanoi:solve(S5)),
    ok.

File: CIS194_intro_haskell/2013_spring/hw1/hanoi.hs
{-# OPTIONS_GHC -Wall #-}

type Peg = [Int]
data PegName = A | B | C deriving Show
data Step = Ab | Ac | Ba | Bc | Ca | Cb deriving Show
type Steps = [Step]
data State = State Peg Peg Peg Steps deriving Show


initState :: Int -> State
initState size = State [1..size] [] [] []


solve :: State -> State
solve state =
    let State a0 _ _ _ = state in
    let State a b c steps = move (length a0) A B C state in
    State a b c (reverse steps)


move :: Int -> PegName -> PegName -> PegName -> State -> State
move numDisks fromPeg toPeg tempPeg state =
    case numDisks of
      0 -> state
      1 -> step fromPeg toPeg state
      2 -> (step tempPeg toPeg) .
           (step fromPeg toPeg) .
           (step fromPeg tempPeg) $ state
      _ -> (move (numDisks - 1) tempPeg toPeg fromPeg) .
           (step fromPeg toPeg) .
           (move (numDisks - 1) fromPeg tempPeg toPeg) $ state


step :: PegName -> PegName -> State -> State
step fromPeg toPeg state =
    let State a b c steps = state in
    case (fromPeg, toPeg) of
      (A, B) -> let (f, t) = place a b in State f t c (Ab : steps)
      (A, C) -> let (f, t) = place a c in State f b t (Ac : steps)
      (B, A) -> let (f, t) = place b a in State t f c (Ba : steps)
      (B, C) -> let (f, t) = place b c in State a f t (Bc : steps)
      (C, A) -> let (f, t) = place c a in State t b f (Ca : steps)
      (C, B) -> let (f, t) = place c b in State a t f (Cb : steps)
      _ -> error "invalid step"


place :: Peg -> Peg -> (Peg, Peg)
place fromPeg toPeg =
    case (fromPeg, toPeg) of
      (fhead : ftail, []) -> (ftail, [fhead])
      (fhead : ftail, thead : _) | fhead < thead -> (ftail, fhead : toPeg)
      _ -> error "invalid place"


main :: IO ()
main =
    putStrLn . show . solve . initState $ 8
